internal/orchestrator/orchestrator.go

```diff
@@ -89,56 +107,115 @@ func (o *Orchestrator) Run(ctx context.Context) {
 		case <-ctx.Done():
 			o.streamBroker.Close()
 			return
+
+		// retrieve and process events sent by controllers.
+		case <-tickerFetchEvents:
+			if o.concurrencyLimit() {
+				continue
+			}
+
+			o.pullEvents(ctx)
+
+		// eventsCh receives events from push based subscriptions, in this case serverservice.
 		case msg := <-eventsCh:
-			o.processMsg(ctx, msg)
+			o.processEvent(ctx, msg)
 		}
 	}
 }
 
-func (o *Orchestrator) processMsg(ctx context.Context, msg events.Message) {
-	otelCtx, span := otel.Tracer(pkgName).Start(ctx, "processMsg")
-	defer span.End()
+func (o *Orchestrator) concurrencyLimit() bool {
+	return int(o.dispatched) >= o.concurrency
+}
 
-	data, err := msg.Data()
+func (o *Orchestrator) pullEvents(ctx context.Context) {
+	// XXX: consider having a separate context for message retrieval
+	msgs, err := o.streamBroker.PullMsg(ctx, 1)
 	if err != nil {
 		o.logger.WithFields(
-			logrus.Fields{"err": err.Error(), "subject": msg.Subject()},
-		).Error("data unpack error")
+			logrus.Fields{"err": err.Error()},
+		).Debug("error fetching work")
+	}
+
+	for _, msg := range msgs {
+		if ctx.Err() != nil || o.concurrencyLimit() {
+			o.eventNak(msg)
+
+			return
+		}
+
+		// spawn msg process handler
+		o.syncWG.Add(1)
+
+		go func(msg events.Message) {
+			defer o.syncWG.Done()
+
+			atomic.AddInt32(&o.dispatched, 1)
+			defer atomic.AddInt32(&o.dispatched, -1)
+
+			o.processEvent(ctx, msg)
+		}(msg)
+	}
+}
+
+func (o *Orchestrator) processEvent(ctx context.Context, event events.Message) {
+	otelCtx, span := otel.Tracer(pkgName).Start(ctx, "processEvent")
+	defer span.End()
+
+	data, err := event.Data()
+	if err != nil {
+		o.ackEvent(event, errors.Wrap(ErrInvalidEvent, "data field error: "+err.Error()))
 
 		return
 	}
 
-	urn, err := msg.SubjectURN(data)
+	urn, err := event.SubjectURN(data)
```

Comment:
Nit: What are examples of the components in the `SubjectURN`? I assume the UUID is the identifier for the server? But where do `namespace` and `prefix` come into play?

Reply:
`namespace` ends up in the [urn](https://github.com/infratographer/x/blob/main/urnx/types.go#L12),  its set to `hollow` in our [configuration](https://github.com/metal-toolbox/architecture/blob/f5b27686e634fed5b0215379f58418728e42227a/messaging.md#event-message-format), as of now its just conforming to what the other teams are following in their NATS messages. `prefix` is used to prefix the Subject when [publishing](https://github.com/metal-toolbox/hollow-toolbox/blob/0bfb6095e5a5566c0874968e1843fd8dddddfef8/events/nats.go#LL219C50-L219C72) the NATS event. Does that answer the query?

Reply:
It does, but it elicits more questions. I guess I don't understand what this URN is for. It doesn't conform to NATs subject rules (and even if it did, I'd object to its use because trying to debug a topic ID labeled with a UUID will be difficult). It seems like the SubjectURN refers to the target of the Condition message, so I'm trying to understand why we use URN instead of any other arbitrary identifier.